### ISSUE TITLE
Client interacts with a Go gRPC server but eventually starts always failing with `Error: 14 UNAVAILABLE: Stream refused by server`

### ISSUE THREAD

**Issue by be9** [grpc/grpc-node#2388]
### Problem description

We are migrating a Node app from legacy `grpc@1.24.11` to `@grpc/grpc-js`. After the upgrade we saw elevated error rates and started digging into what's wrong. One of the errors is `Error: 14 UNAVAILABLE: Stream refused by server` which we could create a clean slate repro for.

As described below, the client works okay until something wrong happens, and all gRPC requests start to fail with `Error: 14 UNAVAILABLE: Stream refused by server`. Restarting the client process helps, but then it hits the error again.

#### Intro

The repro is in https://github.com/be9/grpc-js-go-test. It contains a simple proto service implementation in Go (2 methods which mostly return random strings) and a Node client:

* `make server` starts the server.
* `node client.js <FLAGS>` starts the client.

#### Local server

Running both server and client locally on macOS works without any problems.

Client log:
<details>

```
$ node client.js -p 1 -d 500 --host localhost:50051
2023-03-08T04:47:31.442Z info Starting host=localhost:50051 parallelism=1 delayMs=500 legacyGrpc=false
2023-03-08T04:47:31.443Z info starting loop id=0
2023-03-08T04:47:31.473Z info request complete id=0
2023-03-08T04:47:31.996Z info request complete id=0
2023-03-08T04:47:32.516Z info request complete id=0
2023-03-08T04:47:33.034Z info request complete id=0
2023-03-08T04:47:33.552Z info request complete id=0
2023-03-08T04:47:34.071Z info request complete id=0
2023-03-08T04:47:34.592Z info request complete id=0
2023-03-08T04:47:35.107Z info request complete id=0
2023-03-08T04:47:35.621Z info request complete id=0
2023-03-08T04:47:36.139Z info request complete id=0
2023-03-08T04:47:36.658Z info request complete id=0
2023-03-08T04:47:37.175Z info request complete id=0
2023-03-08T04:47:37.691Z info request complete id=0
2023-03-08T04:47:38.209Z info request complete id=0
2023-03-08T04:47:38.726Z info request complete id=0
2023-03-08T04:47:39.242Z info request complete id=0
2023-03-08T04:47:39.758Z info request complete id=0
2023-03-08T04:47:40.271Z info request complete id=0
2023-03-08T04:47:40.785Z info request complete id=0
2023-03-08T04:47:41.299Z info request complete id=0
2023-03-08T04:47:41.816Z info request complete id=0
2023-03-08T04:47:42.329Z info request complete id=0
2023-03-08T04:47:42.842Z info request complete id=0
2023-03-08T04:47:43.357Z info request complete id=0
2023-03-08T04:47:43.870Z info request complete id=0
```
</details>

Server log:

<details>

```
$ env GRPC_GO_LOG_VERBOSITY_LEVEL=99 GRPC_GO_LOG_SEVERITY_LEVEL=info  make server                                                                                                              2s 742ms
2023/03/08 11:46:58 INFO: [core] [Server #1] Server created
2023/03/08 11:46:58 gRPC server listening at [::]:50051
2023/03/08 11:46:58 INFO: [core] [Server #1 ListenSocket #2] ListenSocket created
2023/03/08 11:46:58 Starting HTTP at :50052
2023/03/08 11:47:31 INFO: [core] CPU time info is unavailable on non-linux environments.
2023/03/08 11:47:31 GetFoos
2023/03/08 11:47:31 GetFoobars [foo_id=0]
2023/03/08 11:47:31 GetFoobars [foo_id=1]
2023/03/08 11:47:31 GetFoobars [foo_id=2]
2023/03/08 11:47:31 GetFoobars [foo_id=3]
2023/03/08 11:47:31 GetFoobars [foo_id=4]
2023/03/08 11:47:31 GetFoobars [foo_id=5]
2023/03/08 11:47:31 GetFoobars [foo_id=6]
2023/03/08 11:47:31 GetFoobars [foo_id=7]
2023/03/08 11:47:31 GetFoobars [foo_id=8]
2023/03/08 11:47:31 GetFoobars [foo_id=9]
2023/03/08 11:47:31 GetFoos
2023/03/08 11:47:31 GetFoobars [foo_id=0]
2023/03/08 11:47:31 GetFoobars [foo_id=1]
2023/03/08 11:47:31 GetFoobars [foo_id=2]
2023/03/08 11:47:31 GetFoobars [foo_id=3]
2023/03/08 11:47:31 GetFoobars [foo_id=4]
2023/03/08 11:47:31 GetFoobars [foo_id=5]
2023/03/08 11:47:31 GetFoobars [foo_id=6]
2023/03/08 11:47:31 GetFoobars [foo_id=7]
2023/03/08 11:47:31 GetFoobars [foo_id=8]
2023/03/08 11:47:31 GetFoobars [foo_id=9]
2023/03/08 11:47:32 GetFoos
2023/03/08 11:47:32 GetFoobars [foo_id=0]
2023/03/08 11:47:32 GetFoobars [foo_id=1]
2023/03/08 11:47:32 GetFoobars [foo_id=2]
2023/03/08 11:47:32 GetFoobars [foo_id=3]
2023/03/08 11:47:32 GetFoobars [foo_id=4]
2023/03/08 11:47:32 GetFoobars [foo_id=5]
2023/03/08 11:47:32 GetFoobars [foo_id=6]
2023/03/08 11:47:32 GetFoobars [foo_id=7]
2023/03/08 11:47:32 GetFoobars [foo_id=8]
2023/03/08 11:47:32 GetFoobars [foo_id=9]
2023/03/08 11:47:33 GetFoos
2023/03/08 11:47:33 GetFoobars [foo_id=0]
2023/03/08 11:47:33 GetFoobars [foo_id=1]
2023/03/08 11:47:33 GetFoobars [foo_id=2]
2023/03/08 11:47:33 GetFoobars [foo_id=3]
2023/03/08 11:47:33 GetFoobars [foo_id=4]
2023/03/08 11:47:33 GetFoobars [foo_id=5]
2023/03/08 11:47:33 GetFoobars [foo_id=6]
2023/03/08 11:47:33 GetFoobars [foo_id=7]
2023/03/08 11:47:33 GetFoobars [foo_id=8]
2023/03/08 11:47:33 GetFoobars [foo_id=9]
2023/03/08 11:47:33 GetFoos
2023/03/08 11:47:33 GetFoobars [foo_id=0]
2023/03/08 11:47:33 GetFoobars [foo_id=1]
2023/03/08 11:47:33 GetFoobars [foo_id=2]
2023/03/08 11:47:33 GetFoobars [foo_id=3]
2023/03/08 11:47:33 GetFoobars [foo_id=4]
2023/03/08 11:47:33 GetFoobars [foo_id=5]
2023/03/08 11:47:33 GetFoobars [foo_id=6]
2023/03/08 11:47:33 GetFoobars [foo_id=7]
2023/03/08 11:47:33 GetFoobars [foo_id=8]
2023/03/08 11:47:33 GetFoobars [foo_id=9]
2023/03/08 11:47:34 GetFoos
2023/03/08 11:47:34 GetFoobars [foo_id=0]
2023/03/08 11:47:34 GetFoobars [foo_id=1]
2023/03/08 11:47:34 GetFoobars [foo_id=2]
2023/03/08 11:47:34 GetFoobars [foo_id=3]
2023/03/08 11:47:34 GetFoobars [foo_id=4]
2023/03/08 11:47:34 GetFoobars [foo_id=5]
2023/03/08 11:47:34 GetFoobars [foo_id=6]
2023/03/08 11:47:34 GetFoobars [foo_id=7]
2023/03/08 11:47:34 GetFoobars [foo_id=8]
2023/03/08 11:47:34 GetFoobars [foo_id=9]
2023/03/08 11:47:34 GetFoos
2023/03/08 11:47:34 GetFoobars [foo_id=0]
2023/03/08 11:47:34 GetFoobars [foo_id=1]
2023/03/08 11:47:34 GetFoobars [foo_id=2]
2023/03/08 11:47:34 GetFoobars [foo_id=3]
2023/03/08 11:47:34 GetFoobars [foo_id=4]
2023/03/08 11:47:34 GetFoobars [foo_id=5]
2023/03/08 11:47:34 GetFoobars [foo_id=6]
2023/03/08 11:47:34 GetFoobars [foo_id=7]
2023/03/08 11:47:34 GetFoobars [foo_id=8]
2023/03/08 11:47:34 GetFoobars [foo_id=9]
2023/03/08 11:47:35 GetFoos
2023/03/08 11:47:35 GetFoobars [foo_id=0]
2023/03/08 11:47:35 GetFoobars [foo_id=1]
2023/03/08 11:47:35 GetFoobars [foo_id=2]
2023/03/08 11:47:35 GetFoobars [foo_id=3]
2023/03/08 11:47:35 GetFoobars [foo_id=4]
2023/03/08 11:47:35 GetFoobars [foo_id=5]
2023/03/08 11:47:35 GetFoobars [foo_id=6]
2023/03/08 11:47:35 GetFoobars [foo_id=7]
2023/03/08 11:47:35 GetFoobars [foo_id=8]
2023/03/08 11:47:35 GetFoobars [foo_id=9]
2023/03/08 11:47:35 GetFoos
2023/03/08 11:47:35 GetFoobars [foo_id=0]
2023/03/08 11:47:35 GetFoobars [foo_id=1]
2023/03/08 11:47:35 GetFoobars [foo_id=2]
2023/03/08 11:47:35 GetFoobars [foo_id=3]
2023/03/08 11:47:35 GetFoobars [foo_id=4]
2023/03/08 11:47:35 GetFoobars [foo_id=5]
2023/03/08 11:47:35 GetFoobars [foo_id=6]
2023/03/08 11:47:35 GetFoobars [foo_id=7]
2023/03/08 11:47:35 GetFoobars [foo_id=8]
2023/03/08 11:47:35 GetFoobars [foo_id=9]
2023/03/08 11:47:36 GetFoos
2023/03/08 11:47:36 GetFoobars [foo_id=0]
2023/03/08 11:47:36 GetFoobars [foo_id=1]
2023/03/08 11:47:36 GetFoobars [foo_id=2]
2023/03/08 11:47:36 GetFoobars [foo_id=3]
2023/03/08 11:47:36 GetFoobars [foo_id=4]
2023/03/08 11:47:36 GetFoobars [foo_id=5]
2023/03/08 11:47:36 GetFoobars [foo_id=6]
2023/03/08 11:47:36 GetFoobars [foo_id=7]
2023/03/08 11:47:36 GetFoobars [foo_id=8]
2023/03/08 11:47:36 GetFoobars [foo_id=9]
2023/03/08 11:47:36 GetFoos
2023/03/08 11:47:36 GetFoobars [foo_id=0]
2023/03/08 11:47:36 GetFoobars [foo_id=1]
2023/03/08 11:47:36 GetFoobars [foo_id=2]
2023/03/08 11:47:36 GetFoobars [foo_id=3]
2023/03/08 11:47:36 GetFoobars [foo_id=4]
2023/03/08 11:47:36 GetFoobars [foo_id=5]
2023/03/08 11:47:36 GetFoobars [foo_id=6]
2023/03/08 11:47:36 GetFoobars [foo_id=7]
2023/03/08 11:47:36 GetFoobars [foo_id=8]
2023/03/08 11:47:36 GetFoobars [foo_id=9]
2023/03/08 11:47:37 GetFoos
2023/03/08 11:47:37 GetFoobars [foo_id=0]
2023/03/08 11:47:37 GetFoobars [foo_id=1]
2023/03/08 11:47:37 GetFoobars [foo_id=2]
2023/03/08 11:47:37 GetFoobars [foo_id=3]
2023/03/08 11:47:37 GetFoobars [foo_id=4]
2023/03/08 11:47:37 GetFoobars [foo_id=5]
2023/03/08 11:47:37 GetFoobars [foo_id=6]
2023/03/08 11:47:37 GetFoobars [foo_id=7]
2023/03/08 11:47:37 GetFoobars [foo_id=8]
2023/03/08 11:47:37 GetFoobars [foo_id=9]
2023/03/08 11:47:37 GetFoos
2023/03/08 11:47:37 GetFoobars [foo_id=0]
2023/03/08 11:47:37 GetFoobars [foo_id=1]
2023/03/08 11:47:37 GetFoobars [foo_id=2]
2023/03/08 11:47:37 GetFoobars [foo_id=3]
2023/03/08 11:47:37 GetFoobars [foo_id=4]
2023/03/08 11:47:37 GetFoobars [foo_id=5]
2023/03/08 11:47:37 GetFoobars [foo_id=6]
2023/03/08 11:47:37 GetFoobars [foo_id=7]
2023/03/08 11:47:37 GetFoobars [foo_id=8]
2023/03/08 11:47:37 GetFoobars [foo_id=9]
2023/03/08 11:47:38 GetFoos
2023/03/08 11:47:38 GetFoobars [foo_id=0]
2023/03/08 11:47:38 GetFoobars [foo_id=1]
2023/03/08 11:47:38 GetFoobars [foo_id=2]
2023/03/08 11:47:38 GetFoobars [foo_id=3]
2023/03/08 11:47:38 GetFoobars [foo_id=4]
2023/03/08 11:47:38 GetFoobars [foo_id=5]
2023/03/08 11:47:38 GetFoobars [foo_id=6]
2023/03/08 11:47:38 GetFoobars [foo_id=7]
2023/03/08 11:47:38 GetFoobars [foo_id=8]
2023/03/08 11:47:38 GetFoobars [foo_id=9]
2023/03/08 11:47:38 GetFoos
2023/03/08 11:47:38 GetFoobars [foo_id=0]
2023/03/08 11:47:38 GetFoobars [foo_id=1]
2023/03/08 11:47:38 GetFoobars [foo_id=2]
2023/03/08 11:47:38 GetFoobars [foo_id=3]
2023/03/08 11:47:38 GetFoobars [foo_id=4]
2023/03/08 11:47:38 GetFoobars [foo_id=5]
2023/03/08 11:47:38 GetFoobars [foo_id=6]
2023/03/08 11:47:38 GetFoobars [foo_id=7]
2023/03/08 11:47:38 GetFoobars [foo_id=8]
2023/03/08 11:47:38 GetFoobars [foo_id=9]
2023/03/08 11:47:39 GetFoos
2023/03/08 11:47:39 GetFoobars [foo_id=0]
2023/03/08 11:47:39 GetFoobars [foo_id=1]
2023/03/08 11:47:39 GetFoobars [foo_id=2]
2023/03/08 11:47:39 GetFoobars [foo_id=3]
2023/03/08 11:47:39 GetFoobars [foo_id=4]
2023/03/08 11:47:39 GetFoobars [foo_id=5]
2023/03/08 11:47:39 GetFoobars [foo_id=6]
2023/03/08 11:47:39 GetFoobars [foo_id=7]
2023/03/08 11:47:39 GetFoobars [foo_id=8]
2023/03/08 11:47:39 GetFoobars [foo_id=9]
2023/03/08 11:47:39 GetFoos
2023/03/08 11:47:39 GetFoobars [foo_id=0]
2023/03/08 11:47:39 GetFoobars [foo_id=1]
2023/03/08 11:47:39 GetFoobars [foo_id=2]
2023/03/08 11:47:39 GetFoobars [foo_id=3]
2023/03/08 11:47:39 GetFoobars [foo_id=4]
2023/03/08 11:47:39 GetFoobars [foo_id=5]
2023/03/08 11:47:39 GetFoobars [foo_id=6]
2023/03/08 11:47:39 GetFoobars [foo_id=7]
2023/03/08 11:47:39 GetFoobars [foo_id=8]
2023/03/08 11:47:39 GetFoobars [foo_id=9]
2023/03/08 11:47:40 GetFoos
2023/03/08 11:47:40 GetFoobars [foo_id=0]
2023/03/08 11:47:40 GetFoobars [foo_id=1]
2023/03/08 11:47:40 GetFoobars [foo_id=2]
2023/03/08 11:47:40 GetFoobars [foo_id=3]
2023/03/08 11:47:40 GetFoobars [foo_id=4]
2023/03/08 11:47:40 GetFoobars [foo_id=5]
2023/03/08 11:47:40 GetFoobars [foo_id=6]
2023/03/08 11:47:40 GetFoobars [foo_id=7]
2023/03/08 11:47:40 GetFoobars [foo_id=8]
2023/03/08 11:47:40 GetFoobars [foo_id=9]
2023/03/08 11:47:40 GetFoos
2023/03/08 11:47:40 GetFoobars [foo_id=0]
2023/03/08 11:47:40 GetFoobars [foo_id=1]
2023/03/08 11:47:40 GetFoobars [foo_id=2]
2023/03/08 11:47:40 GetFoobars [foo_id=3]
2023/03/08 11:47:40 GetFoobars [foo_id=4]
2023/03/08 11:47:40 GetFoobars [foo_id=5]
2023/03/08 11:47:40 GetFoobars [foo_id=6]
2023/03/08 11:47:40 GetFoobars [foo_id=7]
2023/03/08 11:47:40 GetFoobars [foo_id=8]
2023/03/08 11:47:40 GetFoobars [foo_id=9]
2023/03/08 11:47:41 GetFoos
2023/03/08 11:47:41 GetFoobars [foo_id=0]
2023/03/08 11:47:41 GetFoobars [foo_id=1]
2023/03/08 11:47:41 GetFoobars [foo_id=2]
2023/03/08 11:47:41 GetFoobars [foo_id=3]
2023/03/08 11:47:41 GetFoobars [foo_id=4]
2023/03/08 11:47:41 GetFoobars [foo_id=5]
2023/03/08 11:47:41 GetFoobars [foo_id=6]
2023/03/08 11:47:41 GetFoobars [foo_id=7]
2023/03/08 11:47:41 GetFoobars [foo_id=8]
2023/03/08 11:47:41 GetFoobars [foo_id=9]
2023/03/08 11:47:41 INFO: [transport] transport: closing: EOF
2023/03/08 11:47:41 INFO: [transport] transport: loopyWriter exited. Closing connection. Err: transport closed by client
2023/03/08 11:47:41 GetFoos
2023/03/08 11:47:41 GetFoobars [foo_id=0]
2023/03/08 11:47:41 GetFoobars [foo_id=1]
2023/03/08 11:47:41 GetFoobars [foo_id=2]
2023/03/08 11:47:41 GetFoobars [foo_id=3]
2023/03/08 11:47:41 GetFoobars [foo_id=4]
2023/03/08 11:47:41 GetFoobars [foo_id=5]
2023/03/08 11:47:41 GetFoobars [foo_id=6]
2023/03/08 11:47:41 GetFoobars [foo_id=7]
2023/03/08 11:47:41 GetFoobars [foo_id=8]
2023/03/08 11:47:41 GetFoobars [foo_id=9]
2023/03/08 11:47:42 GetFoos
2023/03/08 11:47:42 GetFoobars [foo_id=0]
2023/03/08 11:47:42 GetFoobars [foo_id=1]
2023/03/08 11:47:42 GetFoobars [foo_id=2]
2023/03/08 11:47:42 GetFoobars [foo_id=3]
2023/03/08 11:47:42 GetFoobars [foo_id=4]
2023/03/08 11:47:42 GetFoobars [foo_id=5]
2023/03/08 11:47:42 GetFoobars [foo_id=6]
2023/03/08 11:47:42 GetFoobars [foo_id=7]
2023/03/08 11:47:42 GetFoobars [foo_id=8]
2023/03/08 11:47:42 GetFoobars [foo_id=9]
2023/03/08 11:47:42 GetFoos
2023/03/08 11:47:42 GetFoobars [foo_id=0]
2023/03/08 11:47:42 GetFoobars [foo_id=1]
2023/03/08 11:47:42 GetFoobars [foo_id=2]
2023/03/08 11:47:42 GetFoobars [foo_id=3]
2023/03/08 11:47:42 GetFoobars [foo_id=4]
2023/03/08 11:47:42 GetFoobars [foo_id=5]
2023/03/08 11:47:42 GetFoobars [foo_id=6]
2023/03/08 11:47:42 GetFoobars [foo_id=7]
2023/03/08 11:47:42 GetFoobars [foo_id=8]
2023/03/08 11:47:42 GetFoobars [foo_id=9]
2023/03/08 11:47:43 GetFoos
2023/03/08 11:47:43 GetFoobars [foo_id=0]
2023/03/08 11:47:43 GetFoobars [foo_id=1]
2023/03/08 11:47:43 GetFoobars [foo_id=2]
2023/03/08 11:47:43 GetFoobars [foo_id=3]
2023/03/08 11:47:43 GetFoobars [foo_id=4]
2023/03/08 11:47:43 GetFoobars [foo_id=5]
2023/03/08 11:47:43 GetFoobars [foo_id=6]
2023/03/08 11:47:43 GetFoobars [foo_id=7]
2023/03/08 11:47:43 GetFoobars [foo_id=8]
2023/03/08 11:47:43 GetFoobars [foo_id=9]
2023/03/08 11:47:43 GetFoos
2023/03/08 11:47:43 GetFoobars [foo_id=0]
2023/03/08 11:47:43 GetFoobars [foo_id=1]
2023/03/08 11:47:43 GetFoobars [foo_id=2]
2023/03/08 11:47:43 GetFoobars [foo_id=3]
2023/03/08 11:47:43 GetFoobars [foo_id=4]
2023/03/08 11:47:43 GetFoobars [foo_id=5]
2023/03/08 11:47:43 GetFoobars [foo_id=6]
2023/03/08 11:47:43 GetFoobars [foo_id=7]
2023/03/08 11:47:43 GetFoobars [foo_id=8]
2023/03/08 11:47:43 GetFoobars [foo_id=9]
2023/03/08 11:47:44 INFO: [transport] transport: closing: EOF
2023/03/08 11:47:44 INFO: [transport] transport: loopyWriter exited. Closing connection. Err: transport closed by client
```
</details>

(The server log time is not UTC, but both logs can be correlated by `MM:SS`).

#### Remote server

Now I run the server on a remote Linux machine (ping 180-200ms) hosted in AWS cloud. I've posted the logs to a [separate gist](https://gist.github.com/be9/225f0300ebff44ffb6e97a7c069b6999). If we look at [server.log](https://gist.github.com/be9/225f0300ebff44ffb6e97a7c069b6999#file-server-log), we see that the client has made full 4 rounds of requests, but then no new requests come after `04:51:42`. Client, OTOH, is actively trying to make further requests, but never recovers after seeing a [`HTTP/2 stream closed with code 7`](https://gist.github.com/be9/225f0300ebff44ffb6e97a7c069b6999#file-client-log-L2772) for the first time.

#### Remote server with `grpc@1.24.11`

The legacy library displays no issues. The corresponding logs are [here](https://gist.github.com/be9/a3222560c78bda1e207c468e04f6f895).

(the client command was `GRPC_TRACE=all GRPC_VERBOSITY=debug node client.js -p 1 -d 500 --host $HOST --legacyGrpc`)

### Environment

Local:

- macOS Ventura 13.2.1 (`Darwin odashevskii-mbp 22.3.0 Darwin Kernel Version 22.3.0: Mon Jan 30 20:38:37 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T6000 arm64`).
- Node 18.14.2 installed with `asdf`
- `@grpc/grpc-js@1.8.12`

Remote:

- `Ubuntu 22.04.1 LTS`
- `Linux devenv-odashevskii 5.15.0-1028-aws #32-Ubuntu SMP Mon Jan 9 12:28:07 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux`
- The server is compiled with `go version go1.18 linux/amd64`.

### Additional context

* The client supports parallelism (`-p` flag), but it fails even without it, as shown in the logs. 
* I've also done Wireshark captures and can provide them upon request.

**Comment by murgatroid99** [comment]
I would like to see the Wireshark capture for the failed run. I think it is likely to have some useful information.

**Comment by be9** [comment]
@murgatroid99 attaching. It looks like stream 99 is the first problematic one. 

<img width="1840" alt="Screenshot 2023-03-09 at 07 45 43" src="https://user-images.githubusercontent.com/920/223886554-7aa0f6ff-eb09-43c7-b994-84333c1f264d.png">

[clientside.pcapng.zip](https://github.com/grpc/grpc-node/files/10926553/clientside.pcapng.zip)

**Comment by be9** [comment]
@murgatroid99 the screenshot above actually shows the problem. The server sends `RST_STREAM[97]` (packet 2342). Per [RFC 7540](https://www.rfc-editor.org/rfc/rfc7540):

> open: 
>  ....
>  Either endpoint can send a RST_STREAM frame from this state, causing it to transition immediately to "closed".

.....

> closed:
>  The "closed" state is the terminal state.
>
>  An endpoint MUST NOT send frames other than PRIORITY on a closed stream.  An endpoint that receives any frame other than PRIORITY after receiving a RST_STREAM MUST treat that as a stream error ([Section 5.4.2](https://www.rfc-editor.org/rfc/rfc7540#section-5.4.2)) of type STREAM_CLOSED.

....

> [5.4.2](https://www.rfc-editor.org/rfc/rfc7540#section-5.4.2).  Stream Error Handling
> ....
> To avoid looping, an endpoint MUST NOT send a RST_STREAM in response to a RST_STREAM frame.

However, despite this, the client still sends its own `RST_STREAM[97]` (packet 2346), ignoring the fact that the stream is already in **closed** state. After that, probably, the server goes nuts.

**Comment by murgatroid99** [comment]
I think this is a server bug, and that the right next step is to file an issue with the grpc-go repository. Whatever the client is doing, the server should be able to handle it reasonably, and that would mean either keeping the HTTP2 session in a good state, or sending a GOAWAY.

Also, the Wireshark capture shows that the client almost always sends a single DATA frame with the ES flag set. Specifically on stream 97, it instead sends the HEADERS and a DATA frame in a single TCP packet, and then it does send a separate DATA frame with the ES flag set. But it sends that second frame immediately, before receiving anything from the server, including the RST_STREAM.

---

I would argue that the client is not at all out of spec at all here. The full Stream Error Handling section says (emphasis mine)

> A stream error is an error related to a specific stream **that does not affect processing of other streams**.
>
> An endpoint that detects a stream error sends a [RST_STREAM](https://httpwg.org/specs/rfc9113.html#RST_STREAM) frame ([Section 6.4](https://httpwg.org/specs/rfc9113.html#RST_STREAM)) that contains the stream identifier of the stream where the error occurred. The [RST_STREAM](https://httpwg.org/specs/rfc9113.html#RST_STREAM) frame includes an error code that indicates the type of error.
>
> A [RST_STREAM](https://httpwg.org/specs/rfc9113.html#RST_STREAM) is the last frame that an endpoint can send on a stream. **The peer that sends the [RST_STREAM](https://httpwg.org/specs/rfc9113.html#RST_STREAM) frame MUST be prepared to receive any frames that were sent or enqueued for sending by the remote peer.** These frames can be ignored, except where they modify connection state (such as the state maintained for field section compression ([Section 4.3](https://httpwg.org/specs/rfc9113.html#FieldBlock)) or flow control).
>
> Normally, an endpoint SHOULD NOT send more than one [RST_STREAM](https://httpwg.org/specs/rfc9113.html#RST_STREAM) frame for any stream. **However, an endpoint MAY send additional [RST_STREAM](https://httpwg.org/specs/rfc9113.html#RST_STREAM) frames if it receives frames on a closed stream after more than a round-trip time. This behavior is permitted to deal with misbehaving implementations.**
>
> To avoid looping, an endpoint MUST NOT send a [RST_STREAM](https://httpwg.org/specs/rfc9113.html#RST_STREAM) in response to a [RST_STREAM](https://httpwg.org/specs/rfc9113.html#RST_STREAM) frame.

The first bolded section makes it clear that even if the client did behave incorrectly within a single stream, it should not affect the processing of other streams, as it seems to be doing here.

As the second bolded section shows, there is nothing wrong with an endpoint sending some frames after receiving a RST_STREAM. And the client never sends any other frames after *sending* the RST_STREAM.

And the third bolded section is critical context for the last section, which you quoted. The spec is saying that an endpoint can send additional RST_STREAM frames in response to frames it receives after the stream is closed, but that it should not do that for RST_STREAM frames because that can cause a loop. It is not saying that an endpoint should never send a RST_STREAM frame after receiving one.

**Comment by be9** [comment]
After extensive testing and experimentation we were able to identify the core issue, and it's indeed not in `@grpc/grpc-js`. However, it's not in `grpc-go` either!

The issue was detected in [linkerd-proxy](https://github.com/linkerd/linkerd2-proxy) which was present in our K8S cluster. Linkerd-proxy is written in Rust and uses [h2](https://github.com/hyperium/h2), the HTTP/2 library. H2 had a [bug](https://github.com/hyperium/h2/commit/b0f54d80f24112d198e630e13767501d47ae7290) fixed in version 0.3.16; linkerd-proxy pulled the fix [here](https://github.com/linkerd/linkerd2-proxy/pull/2278), and its recent versions (I've tested 2.12.4) don't have the described problem.

Thanks for your replies here, @murgatroid99! 🙏
